### PR TITLE
Send post_id to the REST API in the ServerSideRender component within the Editor

### LINKED ISSUE #8178: Include 'post_id' in REST API request to render ServerSideRender block
The ServerSideRender REST API endpoint supports a `post_id` argument for setting global `$post` context.

https://github.com/WordPress/gutenberg/blob/4f5295d1e49d28de85802bfd87862d18323c1ce0/lib/class-wp-rest-block-renderer-controller.php#L51-L64

For render callbacks that are dependent on the global `$post` context, it would be helpful to include the `post_id` argument in the REST API request such that the endpoint callback sets up the global `$post`.

https://github.com/WordPress/gutenberg/blob/6928e41c8afd7daa3a709afdda7eee48218473b7/packages/components/src/server-side-render/index.js#L57-L67

It looks like `post_id` was added to the REST API endpoint in a10bfacd4de9ff159807ecf1f05d13323a7eecef but wasn't ever wired in.

Similar to `editor/utils/media-upload`, we'll want to create a `editor/utils/server-side-render/` wrapper that enriches `ServerSideRender` with `getCurrentPostId()`.

### PR DESCRIPTION
**Pull request by alexsanford**
## Description

Fixes https://github.com/WordPress/gutenberg/issues/8178

This PR adds a wrapper around the `ServerSideRender` component within the `@wordpress/editor` package that adds a `post_id` parameter to the REST API call. This ensures the global `$post` object is set properly. Previously, it was only set when rendering to the frontend, but not when rendering for the API call.

This PR is an iteration of this previous PR: https://github.com/WordPress/gutenberg/pull/8270

## How has this been tested?

This was tested locally using a dynamic block that displayed information from the `$post` object, ensuring the post information was rendered properly on the frontend and in the editor using `ServerSideRender`.

An automated test was added to ensure that the parameter is added properly to the URL path.

Also I verified that the current two blocks using `ServerSideRender` (Archives and Latest Comments) still work as expected.

## Types of changes

This change should be a non-breaking enhancement.

That said, I did change the name of the `rendererPathWithAttributes` function to `rendererPath`, and it's technically possible that some 3rd-party code is importing and using that. But I assume that's ok since it's not part of Gutenberg's documented API.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

File: packages/components/src/server-side-render/index.js [hunk 1]
```diff
@@ -20,10 +20,11 @@ import { addQueryArgs } from '@wordpress/url';
 import Placeholder from '../placeholder';
 import Spinner from '../spinner';
 
-export function rendererPathWithAttributes( block, attributes = null ) {
+export function rendererPath( block, attributes = null, urlQueryArgs = null ) {
```
Comment (jorgefilipecosta) [inline packages/components/src/server-side-render/index.js]: Minor: I think for this case we may be able to just do urlQueryArgs = {} on the parameter list, and then ...urlQueryArgs inside addQueryArgs.

File: packages/editor/src/utils/server-side-render/index.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+ * WordPress dependencies.
+ */
+import { ServerSideRender } from '@wordpress/components';
+import { withSelect } from '@wordpress/data';
+
+export default withSelect( ( select ) => {
+	const { getCurrentPostId } = select( 'core/editor' );
+
+	return {
+		urlQueryArgs: {
```
Comment (jorgefilipecosta) [inline packages/editor/src/utils/server-side-render/index.js]: We are ignoring the urlQueryArgs we may be receiving. Maybe we can just check if in urlQueryArgs props this component receives the property post_id is not set and if that's case we add it. urlQueryArgs would default to "{}" so on the most common case we would only add the post_id (equivalent to what we have now).

File: packages/editor/src/utils/index.js [hunk 3]
```diff
@@ -2,5 +2,6 @@
  * Internal dependencies
```
Comment (jorgefilipecosta) [inline packages/editor/src/utils/index.js]: We have some prior art in having a component in @wordpress/editor that adds/enhances props to a component in @worpress/components, e.g., https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/panel-color/index.js and we implemented them in src/components maybe we can also add this one there to be consistent.

**Review by jorgefilipecosta — COMMENTED** [review]
This is looking good 👍 Thank you for iterating on this @alexsanford! I left some comments with possible improvements we can do but from the functional point of view, the tests went great.

File: packages/components/src/server-side-render/index.js (hunk 1 above)
Comment (alexsanford) [reply packages/components/src/server-side-render/index.js]: Good call. Done.

File: packages/editor/src/utils/server-side-render/ReadMe.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# ServerSideRender
```
[the remark this answers is not in the document]
Comment (alexsanford) [reply packages/editor/src/utils/server-side-render/ReadMe.md]: Heh. Turns out I had fixed that in a rebase and forgot to push! 🤦‍♂️ 

Updated the text.

File: packages/editor/src/utils/server-side-render/index.js (hunk 2 above)
Comment (alexsanford) [reply packages/editor/src/utils/server-side-render/index.js]: Good catch! I've updated the implementation for this. It no longer uses `withSelect` but just returns a functional component instead.

File: packages/editor/src/utils/index.js (hunk 3 above)
Comment (alexsanford) [reply packages/editor/src/utils/index.js]: Done!

**Comment by alexsanford** [comment]
@jorgefilipecosta Thanks for the comments! I've addressed them above. Note that since I needed to force push anyway (because of an earlier rebase that didn't get pushed) I went ahead and rebased this onto `master` (no conflicts).

File: packages/components/src/server-side-render/index.js [hunk 5]
```diff
@@ -20,10 +20,11 @@ import { addQueryArgs } from '@wordpress/url';
 import Placeholder from '../placeholder';
 import Spinner from '../spinner';
 
-export function rendererPathWithAttributes( block, attributes = null ) {
+export function rendererPath( block, attributes = null, urlQueryArgs = {} ) {
```
Comment (danielbachhuber) [inline packages/components/src/server-side-render/index.js]: Given this is an exported function, do we need to keep (or deprecate) the original `rendererPathWithAttributes()`?

**Review by danielbachhuber — CHANGES_REQUESTED** [review]
One note to address.

File: packages/components/src/server-side-render/index.js (hunk 5 above)
Comment (danielbachhuber) [reply packages/components/src/server-side-render/index.js]: Phrased another way: maybe we should keep the name as `rendererPathWithAttributes()`, and it simply supports a third `urlQueryArgs` argument.

Comment (jorgefilipecosta) [reply packages/components/src/server-side-render/index.js]: Hi @danielbachhuber, the function rendererPathWithAttributes is exported just for testing but is not exported outside of @wordpress/components as we only export the default export outside of the module (the component) https://github.com/WordPress/gutenberg/blob/0c8481cc72c14499dead222a250d96355fdb4f03/packages/components/src/index.js#L51).
No plugin could have used this function so even if we change the name I don't think deprecations are necessary.

Comment (danielbachhuber) [reply packages/components/src/server-side-render/index.js]: Ah, makes sense. Thanks for clarifying @jorgefilipecosta

**Review by danielbachhuber — APPROVED** [review]

**Review by jorgefilipecosta — APPROVED** [review]
LGTM 👍

**Pull request merged**